### DISCUSSION TITLE
Docker Proxy Cache

### DISCUSSION THREAD

**Discussion by lukashass** [renovatebot/renovate#9004] [Request Help] [answered]
To get around the docker hub rate limit, I'm using the [harbor proxy cache](https://goharbor.io/docs/2.1.0/administration/configure-proxy-cache/)

That results in Dockerfiles containing the following

```Dockerfile
FROM registry.example.com/docker.io/library/node:12.19.1-alpine
```

Can I somehow get renovate (self-hosted) to treat that the same as this?

```Dockerfile
FROM node:12.19.1-alpine
```

**Comment by rarkins** [comment]
Can you describe in more details what you mean? Ideally with an example. Eg do you mean you had to rewrite your Dockerfiles to use your proxy URL?

**Comment by viceice** [comment]
Harbor is not compatible with renovatebot because renovate fetches the full tag list but harbor only returns the already fetched local tags.

So renovate never finds new tags until they are pulled to harbor.

**Comment by lukashass** [comment]
More explanation: I had to rewrite all Dockerfiles to include the harbor registry prefix. That way my CI pulls the images through harbor and I don't run into rate limit problems.

Turns out I haven't even gotten as far as to give renovate credentials to authenticate to my harbor instance :see_no_evil: 
But as per @viceice comment that wouldn't have worked anyway.

That leaves the questions:
- Can I tell renovate to pull the tag list from docker hub directly for images that contain the proxy cache prefix?
- Is pulling the tag list from docker hub rate limited?

**Reply by rarkins** [reply]
You can try setting `registryUrls` to `https://docker.io` but I don't think it will work right now because it's original intent in the Docker datasource was to replace Docker Hub lookups with private ones :)

I am not 100% sure if pulling tags from Docker Hub is rate limited or not - we don't seem to have hit those limits if so.

**Reply by lukashass** [reply]
I finally got around to testing with `registryUrls` and as expected this doesn't work.

I would like to contribute this functionality. My current approach would be to add support for [aliases](https://docs.renovatebot.com/configuration-options/#aliases) to the docker manager.
That way a config like

```json
{
  "aliases": {
    "registry.example.com/docker.io": "docker.io"
  }
}
```
would do the job.

**Reply by rarkins** [reply]
Great, please go ahead

**Reply by lukashass** [reply]
See #10102

**Comment by strowi** [comment]
Hi,

i'm running into a similar issue with [gitlabs dependency-proxy](https://docs.gitlab.com/ee/user/packages/dependency_proxy/).
I think some kind of an alias option would be needed to map the proxy-url to a external registry..

gitlab.XYZ.net/sys/ci/dependency_proxy/containers/docker:18.03 

would map to 

docker.io/docker:18.03

PS: the dockerhub doesn't count HEAD requests but everything else seems to count: https://docs.docker.com/docker-hub/download-rate-limit/

**Reply by rarkins** [reply]
How do you configure `docker` to pull from your gitlab proxy instead of Docker.io?

**Reply by strowi** [reply]
You use "gitlab.XYZ.net/sys/ci/dependency_proxy/containers/docker:18.03" as docker-image where you would normally use "docker:18.03" (including authentication). 

In Gitlab this is not a simple 1:1 pull through proxy:
If you pull "gitlab.XYZ.net/sys/ci/dependency_proxy/containers/xy:12" (with or without authentication, depending if the "sys/ci" group is public or private ) gitlab checks the configurable registry (docker.io is default)  for "xy:12", pulls it and makes it available, even if docker.io is down (in contrast to the "normal" registry-cache provided by docker.io).

**Reply by rarkins** [reply]
So is the problem that Renovate can't access `gitlab.XYZ.net/sys/ci/dependency_proxy/containers/docker:18.03` for that it can't check for new tags/hashes?

**Comment by strowi** [comment]
I did some more testing + searching, renovatebot does use the regular registry-api for listing tags right?
Gitlabs dependency-proxy (probably like most of the other registry-mirror solutions) doesn't seem to expose those endpoints.

With this config:
```json
{
  "hostRules": [
   {
      "hostType": "docker",
      "hostName": "git.xyz.net",
      "username": "strowi",
      "password": "****"
    }
  ]
}
```

.gitlab-ci.yml
```yaml
stages:
  - build

20.10:dind:buildkit:
  stage: build
  image: git.xyz.net/sys/ci/dependency_proxy/containers/docker:18.03
  services:
    - git.xyz.net/sys/ci/dependency_proxy/containers/docker:18.03
```

I'm getting the following 404 on the datasource, probably because of the missing registry-api:
```bash
DEBUG: Found 1 package file(s) (repository=strowi/test2)
 INFO: Dependency extraction complete (repository=strowi/test2)
       "baseBranch": "master",
       "stats": {
         "managers": {"gitlabci": {"fileCount": 1, "depCount": 2}},
         "total": {"fileCount": 1, "depCount": 2}
       }
DEBUG: Datasource 404 (repository=strowi/test2)
       "datasource": "docker",
       "lookupName": "git.xyz.net/sys/ci/dependency_proxy/containers/docker",
       "url": "https://git.xyz.net/users/sign_in"
DEBUG: Failed to look up dependency git.xyz.net/sys/ci/dependency_proxy/containers/docker (repository=strowi/test2, packageFile=.gitlab-ci.yml, dependency=git.xyz.net/sys/ci/dependency_proxy/containers/docker)
```

First thought was, that in this case it would be nice to be able to map `git.XYZ.net/*/*/dependency_proxy/containers/*` to the respective endpoint/image on another registry. That way also harbor and other mirrors (not pull through proxies), could be worked around...

Maybe similar to https://docs.renovatebot.com/configuration-options/#aliases

**Reply by rarkins** [reply]
Proxies often don't support the full API of Docker we need (e.g. listing all tags.. they may only know about the ones they have pulled/cached). I'm not sure why the 404 ended up on /users/sign_in (we don't have that hardcoded anywhere) but it may be a sign of this.

**Reply by strowi** [reply]
Thats what i thought.. oppsite to the upstream docker pull-through cache. 
So there are 2 possibilites if i understand it correctly:

1. don't use a mirror, continue using renovatebot
2. use the mirror, don't use renovatebot for these images.

Unless it would be possible for renovate to detect/map a registry-mirror/path, and check available updates on another registry like docker.io.

**Reply by yangm97** [reply]
Has anybody tried using build args? For instance:

```
FROM node:latest
```

Would become:

```
ARG CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX=docker.io/library
FROM ${CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX}/node:latest
```

Not only would this allow renovate to make a direct api call but also lift the burden from logging into the proxy registry from developers.

**Reply by viceice** [reply]
This isn't supported by renovate.

**Reply by rarkins** [reply]
@viceice which bit specifically don't we support and is it something we could?

**Reply by viceice** [reply]
```Dockerfile
ARG CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX=docker.io/library
FROM ${CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX}/node:latest
```

This is not supported by renovate by default, but can be easily done by using a manual configured regex manager.

**Reply by rarkins** [reply]
Shall we add it as a feature request for the dockerfile manager ?

**Reply by viceice** [reply]
Sure, but we should only support simple args, no interpolation an no nested args

**Reply by rarkins** [reply]
Would not be easy to get to work for the currentValue either

**Reply by lukashass** [reply]
Would this case also be covered by #10118?
`registry` would be `${CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX}` and could thus be replaced using [registryurls](https://docs.renovatebot.com/configuration-options/#registryurls).

**Reply by viceice** [reply]
No, the idea for
```Dockerfile
ARG CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX=docker.io/library
FROM ${CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX}/node:latest
```

is to have

```ini
depName=library/node
currentValue=latest
registryUrl=https://docker.io
```

**Comment by lukashass** [comment]
I just discovered the [regex manager](https://docs.renovatebot.com/modules/manager/regex/) and it seems to be a solution for my problem (at least until the better alternative #10118 is available).

To make renovate work with `FROM registry.example.com/docker.io/library/node:12.19.1-alpine` the same as with `FROM node:12.19.1-alpine` I added the following to my `renovate.json`:

```yaml
  "regexManagers": [
    {
      "fileMatch": ["Dockerfile$"],
      "matchStrings": ["FROM registry\\.example\\.com/docker\\.io/(library/)?(?<depName>.*?):(?<currentValue>.*?)\n"],
      "datasourceTemplate": "docker"
    }
  ]
```

**Comment by cjtim** [comment]
Any ideas for this?

I have Harbor proxy
so `library/golang:123` image

can be pulled via `harbor.company.com/dockerhub/library/golang:123`

so I want to add prefix of `harbor.company.com/dockerhub/` to add image that pulls from dockerhub
How can i do that. Thanks

**Reply by viceice** [reply]
renovate can't use harbor proxy to detect updates because harbor doesn't return the upstream tags list. it only returns already pulled tags.

**Comment by andyinno** [comment]
sorry for commenting on this old thread. Today I found it because I was looking for a solution for the exact same problem.

It turned out that I tried something like the following and it seems to work for me:

```docker
# renovate: datasource=docker depName=docker versioning=ubuntu
ARG DOCKER_BASE_IMAGE=ubuntu:22.04@sha256:965fbcae990b0467ed5657caceaec165018ef44a4d2d46c7cdea80a9dff0d1ea
ARG IMAGE="mygitlabinstance.com:443/path_to_dependency_proxy/containers/$DOCKER_BASE_IMAGE"

FROM ${IMAGE}
```

If I try to build the image, the image is built by fetching the dependency proxy image and renovate looks for the DOCKER_BASE_IMAGE. I will need a confirmation that it is actually working with a freshen image. My renovate is still running while I am writing this.

Anyway. I think that if this is working, a nicer solution would be to pass the dependency proxy url as external arg during the build.

```
docker build -t my_image --build-arg dep_proxy="mydepproxyhere" .
```
 and adding in the dockerfile something like:

```
# renovate: datasource=docker depName=docker versioning=ubuntu
ARG DOCKER_BASE_IMAGE=ubuntu:22.04@sha256:965fbcae990b0467ed5657caceaec165018ef44a4d2d46c7cdea80a9dff0d1ea
ARG dep_proxy

FROM ${dep_proxy}${IMAGE}
```

This would allow to update the dependency proxy without rebuilding the image.

I am not really expert with dockerfile and renovate, therefore I would like to ask for your opinion about this approach.

**Comment by TribuneX** [comment, marked as answer]
I was investigating a similar issue with an image which uses a harbor dockerhub-cache. But fetching of new image tags seems to work now with harbor 2.5.0 and later (see https://github.com/goharbor/harbor/discussions/18187). Just for others finding this discussion. Can the accepted answer be updated?
